Build failure: deno-2.1.4

## Steps To Reproduce

Steps to reproduce the behavior:

1. build deno-2.1.4

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
libtool: compile:  /nix/store/wkadfwbhn9b4kdih88j7izzy20qw79bk-clang-wrapper-19.1.5/bin/cc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -I. -I../include -Iinclude -I../src -Oz -ffunction-sect>
  make[2]: Leaving directory '/private/tmp/nix-build-deno-2.1.4.drv-1/source/target/aarch64-apple-darwin/release/build/libffi-sys-d47a2cef7411a0b3/out/libffi-build/aarch64-apple-darwin'
  make[1]: Leaving directory '/private/tmp/nix-build-deno-2.1.4.drv-1/source/target/aarch64-apple-darwin/release/build/libffi-sys-d47a2cef7411a0b3/out/libffi-build/aarch64-apple-darwin'

  --- stderr
  clang: error: unsupported option '-print-multi-os-directory'
  clang: error: no input files
  ../src/java_raw_api.c:328:46: warning: 'ffi_java_raw_size' is deprecated [-Wdeprecated-declarations]
    328 |   ffi_java_raw *raw = (ffi_java_raw*)alloca (ffi_java_raw_size (cif));
        |                                              ^
  include/ffi.h:326:56: note: 'ffi_java_raw_size' has been explicitly marked deprecated here
    326 | size_t ffi_java_raw_size (ffi_cif *cif) __attribute__((deprecated));
        |                                                        ^
  ../src/java_raw_api.c:331:3: warning: 'ffi_java_ptrarray_to_raw' is deprecated [-Wdeprecated-declarations]
    331 |   ffi_java_ptrarray_to_raw (cif, avalue, raw);
        |   ^
  include/ffi.h:322:93: note: 'ffi_java_ptrarray_to_raw' has been explicitly marked deprecated here
    322 | void ffi_java_ptrarray_to_raw (ffi_cif *cif, void **args, ffi_java_raw *raw) __attribute__((deprecated));
        |                                                                                             ^
  2 warnings generated.
  /private/tmp/nix-build-deno-2.1.4.drv-1/sysv-98dd48.s:27:2: error: invalid CFI advance_loc expression
   .cfi_def_cfa x1, 40;
   ^
  /private/tmp/nix-build-deno-2.1.4.drv-1/sysv-98dd48.s:188:2: error: invalid CFI advance_loc expression
   .cfi_adjust_cfa_offset (8*2 + (8 * 16 + 8 * 8) + 64)
   ^
  make[2]: *** [Makefile:1296: src/aarch64/sysv.lo] Error 1
  make[1]: *** [Makefile:1397: install-recursive] Error 1
  make: *** [Makefile:3158: install] Error 2
  thread 'main' panicked at /private/tmp/nix-build-deno-2.1.4.drv-1/deno-2.1.4-vendor.tar.gz/libffi-sys/build/common.rs:8:5:
```

</details>

## Additional context

<!-- Add any other context about the problem here. -->

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
- system: `"aarch64-darwin"`
 - host os: `Darwin 24.2.0, macOS 15.2`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.25.3`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/73cnf2k7mji95fcflvay15pdw4skhhk5-source`
 
## Notify maintainers
@jk
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/de/deno/package.nix
+++ b/pkgs/by-name/de/deno/package.nix
@@ -7,8 +7,6 @@
   cmake,
   protobuf,
   installShellFiles,
-  apple-sdk_11,
-  darwinMinVersionHook,
   librusty_v8 ? callPackage ./librusty_v8.nix {
     inherit (callPackage ./fetchers.nix { }) fetchLibrustyV8;
   },
@@ -55,14 +53,7 @@ rustPlatform.buildRustPackage rec {
     "--disable-multi-os-directory"
   ];
 
-  buildInputs =
-    [ libffi ]
-    ++ lib.optionals stdenv.hostPlatform.isDarwin [
-      apple-sdk_11
-      # V8 supports 10.15+; binary references `aligned_alloc` directly
-      (darwinMinVersionHook "10.15")
-    ];
-
+  buildInputs = [ libffi ];
   buildAndTestSubdir = "cli";
 
   # work around "error: unknown warning group '-Wunused-but-set-parameter'"

--- a/pkgs/by-name/de/deno/package.nix
+++ b/pkgs/by-name/de/deno/package.nix
@@ -12,6 +12,7 @@
   librusty_v8 ? callPackage ./librusty_v8.nix {
     inherit (callPackage ./fetchers.nix { }) fetchLibrustyV8;
   },
+  libffi,
 }:
 
 let
@@ -33,7 +34,10 @@ rustPlatform.buildRustPackage rec {
   postPatch = ''
     # upstream uses lld on aarch64-darwin for faster builds
     # within nix lld looks for CoreFoundation rather than CoreFoundation.tbd and fails
-    substituteInPlace .cargo/config.toml --replace "-fuse-ld=lld " ""
+    substituteInPlace .cargo/config.toml --replace-fail "-fuse-ld=lld " ""
+
+    # Use patched nixpkgs libffi in order to fix https://github.com/libffi/libffi/pull/857
+    substituteInPlace ext/ffi/Cargo.toml --replace-fail "libffi = \"=3.2.0\"" "libffi = { version = \"3.2.0\", features = [\"system\"] }"
   '';
 
   # uses zlib-ng but can't dynamically link yet
@@ -46,7 +50,12 @@ rustPlatform.buildRustPackage rec {
     installShellFiles
   ];
 
-  buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [
+  configureFlags = lib.optionals stdenv.cc.isClang [
+    # This never worked with clang, but became a hard error recently: https://github.com/llvm/llvm-project/commit/3d5b610c864c8f5980eaa16c22b71ff1cf462fae
+    "--disable-multi-os-directory"
+  ];
+
+  buildInputs = [ libffi ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
     apple-sdk_11
     # V8 supports 10.15+; binary references `aligned_alloc` directly
     (darwinMinVersionHook "10.15")

--- a/pkgs/by-name/de/deno/package.nix
+++ b/pkgs/by-name/de/deno/package.nix
@@ -55,11 +55,13 @@ rustPlatform.buildRustPackage rec {
     "--disable-multi-os-directory"
   ];
 
-  buildInputs = [ libffi ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
-    apple-sdk_11
-    # V8 supports 10.15+; binary references `aligned_alloc` directly
-    (darwinMinVersionHook "10.15")
-  ];
+  buildInputs =
+    [ libffi ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      apple-sdk_11
+      # V8 supports 10.15+; binary references `aligned_alloc` directly
+      (darwinMinVersionHook "10.15")
+    ];
 
   buildAndTestSubdir = "cli";
 
@@ -110,7 +112,10 @@ rustPlatform.buildRustPackage rec {
     '';
     license = licenses.mit;
     mainProgram = "deno";
-    maintainers = with maintainers; [ jk ofalvai ];
+    maintainers = with maintainers; [
+      jk
+      ofalvai
+    ];
     platforms = [
       "x86_64-linux"
       "aarch64-linux"

--- a/pkgs/by-name/de/deno/package.nix
+++ b/pkgs/by-name/de/deno/package.nix
@@ -77,7 +77,7 @@ rustPlatform.buildRustPackage rec {
     find ./target -name libswc_common${stdenv.hostPlatform.extensions.sharedLibrary} -delete
   '';
 
-  postInstall = lib.optionalString (canExecute) ''
+  postInstall = lib.optionalString canExecute ''
     installShellCompletion --cmd deno \
       --bash <($out/bin/deno completions bash) \
       --fish <($out/bin/deno completions fish) \
@@ -85,7 +85,7 @@ rustPlatform.buildRustPackage rec {
   '';
 
   doInstallCheck = canExecute;
-  installCheckPhase = lib.optionalString (canExecute) ''
+  installCheckPhase = lib.optionalString canExecute ''
     runHook preInstallCheck
     $out/bin/deno --help
     $out/bin/deno --version | grep "deno ${version}"
@@ -110,7 +110,7 @@ rustPlatform.buildRustPackage rec {
     '';
     license = licenses.mit;
     mainProgram = "deno";
-    maintainers = with maintainers; [ jk ];
+    maintainers = with maintainers; [ jk ofalvai ];
     platforms = [
       "x86_64-linux"
       "aarch64-linux"

